### PR TITLE
Provide a pre-commit hooks which disallows commit if the code style f…

### PR DESCRIPTION
**Pull request by DeepDiver1975**
…ails

## Description
Committing without having valid php code style in place is pointless.
This pre-commit hook runs make test-php-style and disallows to commit if the code style is wrong.

Please note that this does not automatically install the pre-commit hook. Each and every user has to locally install the hook script

```sh
cd .git/hooks
ln -s ../../build/git-hooks/pre-commit.sh pre-commit
```

## Screenshots (if appropriate):
screenshot from 2018-06-20 08-35-45
screenshot from 2018-06-20 08-36-06

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

**Comment by patrickjahns** [comment]
Please use `.githooks` or `.hooks` in the root folder. Using `build` as a dump all folder is not suitable.

(needs cleaning up from my pov instead of having more permanent files added).

Please add this to the documentation - alternatively instead of committing this into the repository, add it to the developer documentation as best practice?

**Comment by DeepDiver1975** [comment]
> Please add this to the documentation - alternatively instead of committing this into the repository, add it to the developer documentation as best practice?

I think it is more convenient to have the hook script(s) in the git repo and the developer does a simple ln into .git/hooks

**Comment by PVince81** [comment]
how long does the script take ? last time I ran php-cs-fixer locally it seemed to take ages. not nice if this happens for every commit.

if the style checker can be scoped to the commit's changes then ok

**Review by PVince81 — CHANGES_REQUESTED** [review]
style check must only run on actually modified files

**Comment by phil-davis** [comment]
note: php-cs-fixer keeps a cache of stuff it checked previously, so it is much faster after the first run (assuming you didn't change 1000 files in your branch)

**Comment by PVince81** [comment]
not sure if it's my env but currently when I run `make test-php-style` it shows a lot of Fs.
I'm trying to run it to the end again as the previous run had an "out of memory error" despite the allowed 1Gb.

**Comment by PVince81** [comment]
image

**Comment by phil-davis** [comment]
@PVince81 I have a lot less lines of ``.........`` than you. Somehow in your environment it is checking more than just the core files. 

``.php_cs.dist`` should have code in it that builds and exclude list of all the folders in ``apps`` that are not bundled in the core repo.

**Comment by patrickjahns** [comment]
> I think it is more convenient to have the hook script(s) in the git repo and the developer does a simple ln into .git/hooks

talking about convenience - introduce a make rule, that looks for scripts in .githooks and symlinks them into `.git/hooks` ;-)

I leave it up 2 you guys - but before merging - please add it to documentation

**Comment by DeepDiver1975** [comment]
script moved to .githooks

**Comment by patrickjahns** [comment]
How about providing a simple makefile rule to include the githooks:
```
init:
  find .git/hooks -type l -exec rm {} \;
  find .githooks -type f -exec ln -sf ../../{} .git/hooks/ \;
```

Originally by https://www.viget.com/articles/two-ways-to-share-git-hooks-with-your-team/

**Comment by DeepDiver1975** [comment]
> How about providing a simple makefile rule to include the githooks:

nice - wanna commit to this PR right away?

**Comment by PVince81** [comment]
> .php_cs.dist should have code in it that builds an exclude list of all the folders in apps that are not bundled in the core repo.

I tired adding "apps2", "apps3" and "apps-external" to `php_cs.dist` but I still see a forest of `F`.

If you guys agree I can push this here.

**Comment by PVince81** [comment]
I'll use https://github.com/owncloud/core/issues/31863#issuecomment-400575991 to continue investigating why I get errors / out of memory despite exclusion of apps2/apps3

**Comment by PVince81** [comment]
my problems with the style checker have been solved already

carry on with the commit hook ?

File: .githooks/pre-commit.sh [hunk 1]
```diff
@@ -0,0 +1,7 @@
+#!/usr/bin/env bash
+#
+# Installation:
+#  cd .git/hooks
```
Comment (PVince81) [inline .githooks/pre-commit.sh]: extra space before cd

maybe prefix all commands with a "% " as it's universally understood as "run this command"

Comment (DeepDiver1975) [reply .githooks/pre-commit.sh]: It is simply indented by one char - I don't care much how to doc this in here. I guess people will understand that cd and ln are commands to be used on the terminal

Comment (PVince81) [reply .githooks/pre-commit.sh]: fine by me, merge ?

**Pull request closed without merging**